src/Core/Tasks/TaskQueue.cpp

```diff
@@ -178,82 +184,92 @@ void TaskQueue::detectCycles() {
+                               "create a task queue with more threads, or use runTasksInThisThread";
+            return;
+        }
     }
-
     // Add pending dependencies.
     resolveDependencies();
 
     // Do a debug check
     detectCycles();
 
     // Enqueue all tasks with no dependencies.
-    for ( uint t = 0; t < m_tasks.size(); ++t ) {
-        // only queue non null m_tasks
-        if ( m_tasks[t] && m_remainingDependencies[t] == 0 ) { queueTask( TaskId { t } ); }
-    }
+    {
+        wlock lock( m_mutex );
+        for ( uint t = 0; t < m_tasks.size(); ++t ) {
+            // only queue non null m_tasks
 
+            if ( m_tasks[t] && m_remainingDependencies[t] == 0 ) { queueTask( TaskId { t } ); }
+        }
+    }
     // Wake up all threads.
     m_threadNotifier.notify_all();
 }
 
 void TaskQueue::runTasksInThisThread() {
-
-    // lock task queue so no other worker can start working while this thread do the job.
-    std::lock_guard<std::mutex> lock( m_taskQueueMutex );
+    // use local task queue, preventing workers to pickup jobs.
+    std::deque<TaskId> taskQueue;
 
     // Add pending dependencies.
     resolveDependencies();
 
     // Do a debug check
     detectCycles();
-
-    // Enqueue all tasks with no dependencies.
-    for ( uint t = 0; t < m_tasks.size(); ++t ) {
-        // only queue non null m_tasks
-        if ( m_tasks[t] && m_remainingDependencies[t] == 0 ) { queueTask( TaskId { t } ); }
+    {
+        rlock lock( m_mutex );
+
+        // Enqueue all tasks with no dependencies.
+        for ( uint t = 0; t < m_tasks.size(); ++t ) {
+            // only queue non null m_tasks
+            if ( m_tasks[t] && m_remainingDependencies[t] == 0 ) {
+                taskQueue.push_front( TaskId { t } );
+            }
+        }
     }
-    while ( !m_taskQueue.empty() ) {
-        TaskId task { m_taskQueue.back() };
-        m_taskQueue.pop_back();
+    {
+        while ( !taskQueue.empty() ) {
+            TaskId task;
+            {
+                wlock lock( m_mutex );
+                task = taskQueue.back();
+                taskQueue.pop_back();
+            }
+            // Run task
+            m_timerData[task].start    = Utils::Clock::now();
+            m_timerData[task].threadId = 0;
+            m_tasks[task]->process();
+            m_timerData[task].end = Utils::Clock::now();
 
-        // Run task
-        m_timerData[task].start    = Utils::Clock::now();
-        m_timerData[task].threadId = 0;
-        m_tasks[task]->process();
-        m_timerData[task].end = Utils::Clock::now();
-
-        for ( auto t : m_dependencies[task] ) {
-            uint& nDepends = m_remainingDependencies[t];
-            CORE_ASSERT( nDepends > 0, "Inconsistency in dependencies" );
-            --nDepends;
-            if ( nDepends == 0 ) { queueTask( t ); }
+            for ( auto t : m_dependencies[task] ) {
+                wlock lock( m_mutex );
```

Comment:
`taskQueue` is still a local variable in the function (declared at line 216). So, not shared among threads but allocated on the stack in each thread. Are local variables declared in functions executed by several threads shared among those ? I don't think so but perhaps I missed something.

Reply:
My bad, I mixed up between runTask and runTasksInThisThread.
I'll fix it.
The local taskQueue is the most direct solution I find to prevent thread taking tasks during "this thread" run.

When I think about it, the solution might be to have two tasks lists, one for multi-thread, the other for this thread, and they might have inter dependencies. 
Hence GPU tasks in main thread and other tasks could be scheduled by the same taskQueue. Let's postpone for a future PR